File: src/platform_context_graph/resolution/shared_projection/platform_domain.py
"""Helpers for platform-domain shared projection cutover."""

from __future__ import annotations

from typing import Any


def existing_infrastructure_platform_rows(
    session: Any,
    *,
    repo_ids: list[str],
    evidence_source: str,
) -> list[dict[str, str]]:
    """Return existing infrastructure platform edges for targeted repositories."""

    if not repo_ids:
        return []
    return session.run(
        """
        MATCH (repo:Repository)-[rel:PROVISIONS_PLATFORM]->(p:Platform)
        WHERE repo.id IN $repo_ids
          AND rel.evidence_source = $evidence_source
        RETURN repo.id as repo_id,
               p.id as existing_platform_id
        ORDER BY repo.id, p.id
        """,
        repo_ids=repo_ids,
        evidence_source=evidence_source,
    ).data()


def build_platform_infra_intent_rows(
    *,
    descriptor_rows: list[dict[str, object]],
    existing_rows: list[dict[str, str]],
) -> list[dict[str, object]]:
    """Return authoritative platform-domain upsert and retract intent rows."""

    desired_pairs = {
        (
            str(row.get("repo_id") or ""),
            str(row.get("platform_id") or ""),
        )
        for row in descriptor_rows
    }
    intent_rows = [dict(row, action="upsert") for row in descriptor_rows]
    for row in existing_rows:
        pair = (
            str(row.get("repo_id") or ""),
            str(row.get("existing_platform_id") or ""),
        )
        if not pair[0] or not pair[1] or pair in desired_pairs:
            continue
        intent_rows.append(
            {
                "repo_id": pair[0],
                "platform_id": pair[1],
                "action": "retract",
            }
        )
    return intent_rows


def shared_platform_projection_metrics(
    *,
    intent_rows: list[dict[str, object]],
    projection_context_by_repo_id: dict[str, dict[str, str]] | None,
) -> dict[str, object]:
    """Return completion-fencing metadata for authoritative platform cutover."""

    touched_repositories = {
        str(row.get("repo_id") or "") for row in intent_rows if row.get("repo_id")
    }
    if not touched_repositories or not projection_context_by_repo_id:
        return {}
    generation_ids = {
        projection_context_by_repo_id[repo_id]["generation_id"]
        for repo_id in touched_repositories
        if repo_id in projection_context_by_repo_id
    }
    return {
        "authoritative_domains": ["platform_infra"],
        "accepted_generation_id": (
            next(iter(generation_ids)) if len(generation_ids) == 1 else None
        ),
        "intent_count": len(intent_rows),
    }

File: go/internal/reducer/platform_domain.go
package reducer

// PlatformIntentAction represents an upsert or retract action.
const (
	IntentActionUpsert  = "upsert"
	IntentActionRetract = "retract"
)

// ExistingPlatformEdge represents one existing PROVISIONS_PLATFORM edge.
type ExistingPlatformEdge struct {
	RepoID             string
	ExistingPlatformID string
}

// PlatformInfraIntentRow is one authoritative platform-domain intent row with action.
type PlatformInfraIntentRow struct {
	RepoID     string
	PlatformID string
	Action     string // "upsert" or "retract"
	// Additional descriptor fields from the original row.
	Extra map[string]any
}

// BuildPlatformInfraIntentRows diffs desired descriptor rows against existing
// platform edges and returns authoritative upsert and retract intent rows.
func BuildPlatformInfraIntentRows(
	descriptorRows []map[string]any,
	existingRows []ExistingPlatformEdge,
) []PlatformInfraIntentRow {
	// Build set of desired (repo_id, platform_id) pairs from descriptor rows.
	desiredPairs := make(map[[2]string]struct{})
	for _, row := range descriptorRows {
		repoID := anyToString(row["repo_id"])
		platformID := anyToString(row["platform_id"])
		if repoID != "" && platformID != "" {
			desiredPairs[[2]string{repoID, platformID}] = struct{}{}
		}
	}

	// Start with all descriptor rows marked as upserts.
	intentRows := make([]PlatformInfraIntentRow, 0, len(descriptorRows)+len(existingRows))
	for _, row := range descriptorRows {
		repoID := anyToString(row["repo_id"])
		platformID := anyToString(row["platform_id"])

		// Copy all fields except repo_id and platform_id into Extra.
		extra := make(map[string]any)
		for k, v := range row {
			if k != "repo_id" && k != "platform_id" {
				extra[k] = v
			}
		}

		intentRows = append(intentRows, PlatformInfraIntentRow{
			RepoID:     repoID,
			PlatformID: platformID,
			Action:     IntentActionUpsert,
			Extra:      extra,
		})
	}

	// Add retract rows for existing edges not in desired pairs.
	for _, row := range existingRows {
		repoID := row.RepoID
		platformID := row.ExistingPlatformID

		// Skip empty pairs.
		if repoID == "" || platformID == "" {
			continue
		}

		// Skip if this pair is in the desired set.
		pair := [2]string{repoID, platformID}
		if _, ok := desiredPairs[pair]; ok {
			continue
		}

		// Add retract row.
		intentRows = append(intentRows, PlatformInfraIntentRow{
			RepoID:     repoID,
			PlatformID: platformID,
			Action:     IntentActionRetract,
			Extra:      make(map[string]any),
		})
	}

	return intentRows
}

// SharedPlatformProjectionMetrics returns completion-fencing metadata for
// authoritative platform cutover.
func SharedPlatformProjectionMetrics(
	intentRows []PlatformInfraIntentRow,
	contextByRepoID map[string]ProjectionContext,
) map[string]any {
	// Collect touched repository IDs from intent rows.
	touchedRepos := make(map[string]struct{})
	for _, row := range intentRows {
		if row.RepoID != "" {
			touchedRepos[row.RepoID] = struct{}{}
		}
	}

	// Return empty if no touched repositories or no context.
	if len(touchedRepos) == 0 || contextByRepoID == nil {
		return map[string]any{}
	}

	// Collect unique generation IDs for touched repositories.
	generationIDs := make(map[string]struct{})
	for repoID := range touchedRepos {
		if ctx, ok := contextByRepoID[repoID]; ok {
			if ctx.GenerationID != "" {
				generationIDs[ctx.GenerationID] = struct{}{}
			}
		}
	}

	// Build result map.
	result := map[string]any{
		"authoritative_domains": []string{"platform_infra"},
		"intent_count":          len(intentRows),
	}

	// Set accepted_generation_id only if exactly one unique generation ID.
	if len(generationIDs) == 1 {
		for genID := range generationIDs {
			result["accepted_generation_id"] = genID
			break
		}
	} else {
		result["accepted_generation_id"] = nil
	}

	return result
}
